core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,377 @@
+      }
+    }
+  }
+
+  /**
+   * Reads the trust certificates specified in the path location, and update the key store if the
+   * modified time has changed since last read.
+   *
+   * @param trustCertFile  the file on disk holding the trust certificates
+   * @param oldTime the time when the trust file is modified during last execution
+   * @return oldTime if failed or the modified time is not changed, otherwise the new modified time
+   */
+  private long readAndUpdate(File trustCertFile, long oldTime)
+      throws CertificateException, IOException, KeyStoreException, NoSuchAlgorithmException {
+    long newTime = trustCertFile.lastModified();
+    if (newTime != oldTime) {
+      FileInputStream inputStream = new FileInputStream(trustCertFile);
+      try {
+        X509Certificate[] certificates = CertificateUtils.getX509Certificates(inputStream);
+        updateTrustCredentials(certificates);
+        return newTime;
+      } finally {
+        inputStream.close();
+      }
+    }
+    return oldTime;
+  }
+
+  // Mainly used to avoid throwing IO Exceptions in java.io.Closeable.
+  public interface Closeable extends java.io.Closeable {
+    @Override public void close();
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  // The verification mode when authenticating the peer certificate.
+  public enum Verification {
+    // This is the DEFAULT and RECOMMENDED mode for most applications.
+    // Setting this on the client side will do the certificate and hostname verification, while
+    // setting this on the server side will only do the certificate verification.
+    CertificateAndHostNameVerification,
+    // This SHOULD be chosen only when you know what the implication this will bring, and have a
+    // basic understanding about TLS.
+    // It SHOULD be accompanied with proper additional peer identity checks set through
+    // {@code PeerVerifier}(nit: why this @code not working?). Failing to do so will leave
+    // applications to MITM attack.
+    // Also note that this will only take effect if the underlying SDK implementation invokes
+    // checkClientTrusted/checkServerTrusted with the {@code SSLEngine} parameter while doing
+    // verification.
+    // Setting this on either side will only do the certificate verification.
+    CertificateOnlyVerification,
+    // Setting is very DANGEROUS. Please try to avoid this in a real production environment, unless
+    // you are a super advanced user intended to re-implement the whole verification logic on your
+    // own. A secure verification might include:
+    // 1. proper verification on the peer certificate chain
+    // 2. proper checks on the identity of the peer certificate
+    InsecurelySkipAllVerification,
+  }
+
+  // Additional custom peer verification check.
+  // It will be used when checkClientTrusted/checkServerTrusted is called without
```

Comment:
I think this sort of conditional calling will be annoying for implementors and ask for security holes. Either we should pass `null` as the Socket/SSLEngine or just throw "not trusted" if we get called without them. I like the idea of just saying "not trust" if there is no Socket/SSLEngine; it shouldn't really happen these days anyway (it is an old API).